### PR TITLE
[feature] [backward-incompat] [null support # 2] Preserve null literal information in literal context and literal transform

### PR DESCRIPTION
**Pull request by 61yao**
This PR preserves the type information of literal in LiteralContext and LiteralTransform.

This means numerical literals will be passed down from broker either as a long or double except for big decimal. 

String version of numerical values are not longer accepted. For example, "123" won't be treated as a numerical anymore.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java [hunk 1]
```diff
@@ -115,10 +171,7 @@ public boolean equals(Object o) {
...
   public String toString() {
-    if (_value == null) {
-      return "null";
-    }
     // TODO: print out the type.
-    return '\'' + _value.toString() + '\'';
+    return '\'' + String.valueOf(_value) + '\'';
   }
-}
+}
```
Comment (zhtaoxiang) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:177]: nit: new line

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunctionTest.java [hunk 2]
```diff
@@ -90,51 +88,53 @@ public void testCasePriorityObserved(String column, int threshold1, int threshol
...
+          getExpectedLongResults(DOUBLE_SV_COLUMN, functionType));
+      testCaseQueryWithLongResults(String.format("%s(%s, %s)", functionType.getName(), STRING_SV_COLUMN,
               String.format("'%s'", _stringSVValues[INDEX_TO_COMPARE])),
-          getExpectedIntResults(STRING_SV_COLUMN, functionType));
+          getExpectedLongResults(STRING_SV_COLUMN, functionType));
     }
   }
 
   @Test
   public void testCaseTransformFunctionWithFloatResults() {
```
Comment (zhtaoxiang) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunctionTest.java:116]: nit: testCaseTransformFunctionWithFloatResults -> testCaseTransformFunctionWithDoubleResults

File: pinot-core/src/test/java/org/apache/pinot/common/request/context/LiteralContextTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    Assert.assertEquals(LiteralContext.inferLiteralDataTypeAndValue("1.2").getLeft(), FieldSpec.DataType.STRING);
+    Assert.assertEquals(LiteralContext.inferLiteralDataTypeAndValue("41241241.2412").getLeft(),
+        FieldSpec.DataType.STRING);
+    // Only big decimal is allowed to pass in as string for numerical type.
+    Assert.assertEquals(LiteralContext.inferLiteralDataTypeAndValue("1.7976931348623159e+308").getLeft(),
+        FieldSpec.DataType.BIG_DECIMAL);
+    Assert.assertEquals(LiteralContext.inferLiteralDataTypeAndValue("2020-02-02 20:20:20.20").getLeft(),
+        FieldSpec.DataType.TIMESTAMP);
+  }
+}
```
[the remark this answers is not in the document]
Comment (61yao) [reply pinot-core/src/test/java/org/apache/pinot/common/request/context/LiteralContextTest.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 1 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java]: done

**Comment by gortiz** [comment]
These changes already include the ones in https://github.com/apache/pinot/pull/10376, right?

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java [hunk 4]
```diff
@@ -37,68 +45,116 @@ public class LiteralContext {
...
+        break;
+      case STRING_VALUE:
+        Pair<FieldSpec.DataType, Object> typeAndValue = inferLiteralDataTypeAndValue(literal.getFieldValue().toString());
+        _type = typeAndValue.getLeft();
+        _value = typeAndValue.getRight();
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported data type:" + literal.getSetField());
+    }
+    _bigDecimalValue = getBigDecimalValue(_type, _value);
```
Comment (gortiz) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:118]: It is not clear to me what is the difference between `_value` and `_bigDecimalValue` in each case.

* If literal.getSetField is `DOUBLE_VALUE` or `LONG_VALUE`, `_value` is either a `Double` or a `Long`.
* If literal.getSetField is `STRING_VALUE`:
     * And the literal is an actual number (like `"123"` or `"1.3"`), then `_value` is the more specific implementation `Number` possible.
     * And the literal is not an actual number:
             * If it is a boolean, `_value` is a String like `"true"` or `"false"` and `_bigDecimanValue` is `1` or `0`.
             * If it is a timestamp, `_value` is a String (why?) like "21312312" and `_bigDecimanValue` is the representation as a BigDecimal.

Then `_bigDecimalValue` is only used in `getXValue` where `X` is `Int`, `Double` or `BigDecimal`.

For the context, `BigDecimal` instances are quite large. They contain several attributes and therefore they consume a significant amount of memory. I guess normal queries should not have tons of instances of LiteralContext, but we will receive degenerated queries like `where X in [list with hundreds/thousand of elements]` or `where X = L1 or X = L2 or ... X = L1000`. I had to deal with these cases in other databases and suddenly inefficiencies that looked innocent imply OOMs.

Therefore I would strongly recommend to try to not use `BigDecimal` if not needed. In this case:
- Most of the times we could try first to check whether _value is instance of Number, in which case we can just call `intValue` or `doubleValue`.
- Substitute final `_bigDecimalValue` with mutable `_numberValue` (or final `AtomicReference`). This attribute will not be instantiated unless `getBigDecimalValue` is called or in case `getIntValue` or `getDoubleValue` is called in a literal of non numeric type.
- Therefore `getXValue` should check whether the value is null and calculate the actual value if it is not.
- When the value is calculated, ideally we should not use a `_bigDecimalValue` unless we are actually calling `getBigDecimalValue`, in which case we have no other option.

By doing that we may reduce the amount of memory used in degenerated queries like the ones I listed above.

I'm assuming these `getXValue` are going to be called in the hotpath. In case they don't I wouldn't even care trying to cache the value.
- `getBigDecimalValue`

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java [hunk 5]
```diff
@@ -37,68 +45,116 @@ public class LiteralContext {
...
+    _bigDecimalValue = getBigDecimalValue(type, value);
   }
 
   @Override
   public int hashCode() {
-    if (_value == null) {
-      return 31 * _type.hashCode();
-    }
-    return 31 * _value.hashCode();
+    return 31 * Objects.hashCode(_value) + Objects.hashCode(_type);
```
Comment (gortiz) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:157]: Why this change in the hash code?

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/ExpressionContext.java [hunk 6]
```diff
@@ -65,15 +65,6 @@ private ExpressionContext(Type type, String identifier, FunctionContext function
     _literal = literal;
   }
 
-  // TODO: Refactor all of the usage for getLiteralString.
-  @Deprecated
-  public String getLiteralString() {
```
Comment (gortiz) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/ExpressionContext.java:70]: In order to reduce the number of changes... why not just change this code to call `_literal.getStringValue()`?

Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/ExpressionContext.java:70]: because we want caller to explicitly specify what type they want to use. so they have to use getLiteral().getTypeLiteral()

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 4 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:118]: _bigDecimalValue is a cached representation for numerical types. We did the same cache in LiteralTransform func so we don't need to calculate numerical representation of value multiple times.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 5 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:157]: because originally we are not table to tell "123" and 123. After the type reservation, "123" and 123 are different. thus we add the type encoding into the hashcode.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 4 above)
Comment (gortiz) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:118]: Correct me if I'm wrong, but `LiteralContext.getXValue` is only called once from `LiteralTransform` constructor/init, right? Then we don't need to catch it.

I think `LiteralTransform` shouldn't catch all the values _just in case_ someone needs them. I'm not saying we should change `LiteralTransform` asap, but it would be great to do not replicate antipatterns

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 5 above)
Comment (gortiz) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:157]: > because originally we are not table to tell "123" and 123. After the type reservation

Yeah, but I mean... why do we add the hash of types? That isn't usual when producing hashes. In fact if you do `return Objects.hashCode(_value, _type)` it will produce better hashCode (at the cost of creating a short living array).

**Review by gortiz — APPROVED** [review]
I would change the HashCode function, but apart from that, it looks fine to me.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 5 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java]: good point. changed to Objects.hash(...)

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java (hunk 4 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java:118]: not really. for example, LiteralContext.getIntValue() gets called in HistogramAggregationFunction. that's why I cache it.

**Comment by 61yao** [comment]
> I would change the HashCode function, but apart from that, it looks fine to me.
Thanks a lot for the review!

**Review by KKcorps — APPROVED** [review]
